### DISCUSSION TITLE
Failed running pyinstaller for universal2 arch. Some system .so dependencies are not compatible

### DISCUSSION THREAD

**Discussion by iradization** [pyinstaller/pyinstaller#7740] [Features & Development]
I've set target_arch="universal2" to my pyinstaller command to build a macho fat file for running on both Intel and arm64 macOS available architectures.

However, some of the dependancies are python system .so libraries built for current arch only (in my case it's arm64)

This leads to the following error when trying to build the executable:

```
    raise IncompatibleBinaryArchError(f"{display_name} is not a fat binary!")
PyInstaller.utils.osx.IncompatibleBinaryArchError:
 /opt/homebrew/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/lib-dynload/_struct.cpython-310-darwin.so is not a fat binary!
```

So apparently _struct.cpython-310-darwin.so is built for arm64 only... Is there any way to build a FAT version of my project ?

where can I get builtin python framework that compiled as FAT ?

Is there a way to auto-download the right library compiled for the desired arch, upon demand ?

thanks !

**Comment by rokm** [comment]
You cannot use Homebrew python to make `universal2` binaries, because its shared library and built-in extensions are all single-arch. [Python.org](https://www.python.org/downloads/macos/) offers `universal2` python builds. 

But even with `universal2` python environment, some packages may provide single-arch wheels, which will prevent you from building a `universal2` application. In that case, you'll either need to manually stitch the offending shared libraries/extensions together by manually downloading both wheel variants and using `lipo` (and then replace the single-arch file with stitched fat one), or figure out how to locally compile that package in `universal2` mode.
